### ISSUE TITLE
SQL History

### ISSUE THREAD

**Issue by gwired2015** [microsoft/azuredatastudio#95]
You have task history.  Why not show the SQL History of executed SQL Statements?  Store this locally as JSON, XML, Text etc.  Instead of saving SQL files it is much easier to have a list of them.  Store the date of execution as well.  Must be searchable.

**Comment by kevcunnane** [comment]
Can you clarify what exact data you'd like? So far I see
- Query Text
  - Would this be all text executed (e.g. the file contents if you hit `Execute`) or would each batch (separated by `GO`) be stored separately?
- Date & Time executed

What about result information, for example errors, number of results? Any other types of info you need recorded?

Overall this is an interesting feature to think about. We'd definitely need to scope it out and understand what most users wanted from this, and ideally hear back to help prioritize. I also think this could be a great community extension as we move towards opening up extensibility in the longer term.

**Comment by gwired2015** [comment]
Well, I have two applications that have had this feature.

SQL Recall in TOAD for SQL
SSMS Tools SQL History

It stores the executed query string.   Not what is in the window.

If I have a bunch of SQL and I execute one piece or have one SQL that I'm
adjusting it stores each executed bit as a line entry.

And it stores it as just the text I executed.

I use this a lot because I will typically be working on diagnosing a
problem and all the queries for that diagnosis will be run in a short
period of time.  I love SSMS but would defer to TOAD only for that history
feature.

Hope that clarifies.

Garrett

**Comment by tomburgin** [comment]
Things I would want if possible:
- File contents when executed (breaking down per batch doesn't give you the whole picture of what you were doing at the time)
- Server and database the query was run on
- Date and time executed

**Comment by gwired2015** [comment]
I'm sure the other info would be useful for some.

SQL Text
Last Date / Time Execution / #results /Execution time.

**Comment by martincroft** [comment]
The ability to clear the task history doesn't seem to be a feature which would be very useful, I've tagged on here but didn't know if it should be a new request.  

The task history doesn't show backups running if it ran from a script. When running a restore it does not show any indicator of progress ( like you would see in a script % completed).

**Comment by Rich-Ott** [comment]
I find that storing the following information has been most useful:

* Query Text
  * This stores the full text of the executed query
* Date/Time that the query was executed
* Server Name the query was run on
* Database Name the query was run against

The query history also needs to be searchable and filterable by date range.

**Comment by gwired2015** [comment]
Adding on here:
For the UI, would like to see this on the History SideBar.
Task History and Execution History separated similar to the way that Each Server is separated. 
Another way to get to same info would be context menu on each server.
Within History, it should be searchable and search the text of the executions.  
I could see myself using it both ways as a quick way on the Server Sidebar to see history and in the History flyout.  
Another nice addition would be to make the Sidebar dockable to the right side.   Or any "sub" sidebar(meaning if the sidebar has multiple callouts like each server group in servers) so I can permanently keep items I commonly open on the right.

**Comment by AndrewDP23** [comment]
The lack of this feature stops me from using Azure Data Studio. I use an add-on to SSMS to auto-save executed queries. Forgetting to save an ad-hoc SQL query then having it lost is concerning.

**Comment by SQLvariant** [comment]
Once you have the SQL History available, I would like the ability to pick and choose multiple (but not all) commands in the SQL History and 'curate' them into a single SQL Notebook.  That way it's easy to repeat the steps again in the future.  Basically, each command I pick out of the SQL History should go into it's own cell in the SQL Notebook.

Even better if you're able to include metadata about how long the command originally took to run when you copy it over.

**Comment by dtrue** [comment]
Any way this SQL History could be enhanced to not be just session based.  It would be great to have this feature even after opening closing sessions.  If performance is a potential issue just have a setting for max query history results or something.

**Comment by Charles-Gagnon** [comment]
I'm going to close this out since as of the October 2019 release we have added the Query History extension. Try it out and let us know what you think!

https://docs.microsoft.com/en-us/sql/azure-data-studio/release-notes-azure-data-studio?view=sql-server-ver15#october-2019

@dtrue I've opened a feature request issue for your request. Thanks for the feedback!

https://github.com/microsoft/azuredatastudio/issues/8243
